<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.551">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Lucas Kook - Curriculum Vitae</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Lucas Kook</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./publications.html"> 
<span class="menu-text">Publications</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./software.html"> 
<span class="menu-text">Software</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./teaching.html"> 
<span class="menu-text">Teaching</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./reviewing.html"> 
<span class="menu-text">Reviewing</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./talks.html"> 
<span class="menu-text">Talks</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./contact.html"> 
<span class="menu-text">Contact</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#education" id="toc-education" class="nav-link active" data-scroll-target="#education">Education</a></li>
  <li><a href="#work-experience" id="toc-work-experience" class="nav-link" data-scroll-target="#work-experience">Work experience</a></li>
  <li><a href="#teaching-experience" id="toc-teaching-experience" class="nav-link" data-scroll-target="#teaching-experience">Teaching Experience</a>
  <ul class="collapse">
  <li><a href="#current-teaching-responsibilities" id="toc-current-teaching-responsibilities" class="nav-link" data-scroll-target="#current-teaching-responsibilities">Current teaching responsibilities</a></li>
  <li><a href="#student-supervision" id="toc-student-supervision" class="nav-link" data-scroll-target="#student-supervision">Student supervision</a></li>
  <li><a href="#teaching-experience-1" id="toc-teaching-experience-1" class="nav-link" data-scroll-target="#teaching-experience-1">Teaching experience</a></li>
  </ul></li>
  <li><a href="#reviewing" id="toc-reviewing" class="nav-link" data-scroll-target="#reviewing">Reviewing</a>
  <ul class="collapse">
  <li><a href="#editorial-responsibilities" id="toc-editorial-responsibilities" class="nav-link" data-scroll-target="#editorial-responsibilities">Editorial responsibilities</a></li>
  <li><a href="#peer-review" id="toc-peer-review" class="nav-link" data-scroll-target="#peer-review">Peer review</a></li>
  </ul></li>
  <li><a href="#awards-and-scholarships" id="toc-awards-and-scholarships" class="nav-link" data-scroll-target="#awards-and-scholarships">Awards and Scholarships</a></li>
  <li><a href="#languages" id="toc-languages" class="nav-link" data-scroll-target="#languages">Languages</a></li>
  <li><a href="#software" id="toc-software" class="nav-link" data-scroll-target="#software">Software</a>
  <ul class="collapse">
  <li><a href="#maintained-r-packages" id="toc-maintained-r-packages" class="nav-link" data-scroll-target="#maintained-r-packages">Maintained R packages</a></li>
  <li><a href="#contributed-r-packages" id="toc-contributed-r-packages" class="nav-link" data-scroll-target="#contributed-r-packages">Contributed R packages</a></li>
  <li><a href="#contributed-python-packages" id="toc-contributed-python-packages" class="nav-link" data-scroll-target="#contributed-python-packages">Contributed Python packages</a></li>
  </ul></li>
  <li><a href="#publications" id="toc-publications" class="nav-link" data-scroll-target="#publications">Publications</a>
  <ul class="collapse">
  <li><a href="#preprints" id="toc-preprints" class="nav-link" data-scroll-target="#preprints">Preprints</a></li>
  <li><a href="#methodology" id="toc-methodology" class="nav-link" data-scroll-target="#methodology">Methodology</a></li>
  <li><a href="#biomedicine" id="toc-biomedicine" class="nav-link" data-scroll-target="#biomedicine">Biomedicine</a></li>
  </ul></li>
  <li><a href="#talks" id="toc-talks" class="nav-link" data-scroll-target="#talks">Talks</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Curriculum Vitae</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="education" class="level2">
<h2 class="anchored" data-anchor-id="education">Education</h2>
<ul>
<li>Ph.D in Biostatistics, University of Zurich, 03/2020–02/2023</li>
<li>M.S. in Biostatistics, University of Zurich, Switzerland, 09/2018–02/2020</li>
<li>B.S. in Molecular Medicine, University of Freiburg, Germany, 10/2015–06/2018</li>
</ul>
</section>
<section id="work-experience" class="level2">
<h2 class="anchored" data-anchor-id="work-experience">Work experience</h2>
<ul>
<li>03/2024–present Assistant professor of Statistics
<ul>
<li>Vienna University of Economics and Business</li>
</ul></li>
<li>03/2023–02/2024 Postdoctoral researcher
<ul>
<li>University of Copenhagen</li>
<li>Collaborators: Prof.&nbsp;Niklas Pfister, Prof.&nbsp;Jonas Peters</li>
<li>Funded by a “Postdoc.Mobility” grant from the Swiss National Science Foundation on <a href="https://data.snf.ch/grants/grant/214457">Distributional Instrumental Variables Estimation (DIVE)</a></li>
</ul></li>
<li>04/2022–07/2022 Visiting PhD student
<ul>
<li>University of Copenhagen</li>
<li>Supervisor: Prof.&nbsp;Jonas Peters</li>
</ul></li>
<li>03/2020–02/2023: PhD candidate in Biostatistics
<ul>
<li>University of Zurich, Zurich University of Applied Sciences</li>
<li>Supervisors: Prof.&nbsp;Beate Sick, Prof.&nbsp;Torsten Hothorn</li>
</ul></li>
<li>11/2018–08/2019: Research assistant
<ul>
<li>Functional Genomics Center Zurich, University of Zurich, ETH Zurich</li>
<li>Supervisor: Dr.&nbsp;Witold Wolski</li>
</ul></li>
<li>01/2017–08/2018: Research assistant
<ul>
<li>University of Freiburg</li>
<li>Supervisor: Prof.&nbsp;Oliver Schilling</li>
</ul></li>
</ul>
</section>
<section id="teaching-experience" class="level2">
<h2 class="anchored" data-anchor-id="teaching-experience">Teaching Experience</h2>
<section id="current-teaching-responsibilities" class="level3">
<h3 class="anchored" data-anchor-id="current-teaching-responsibilities">Current teaching responsibilities</h3>
<ul>
<li><p>Data Analytics (WU Vienna, Bachelor level, 60 students, SS24, WS24, SS25, WS25)</p></li>
<li><p>Statistics (WU Vienna, Bachelor level, 180 students, SS24, WS24, SS25, WS25)</p></li>
</ul>
</section>
<section id="student-supervision" class="level3">
<h3 class="anchored" data-anchor-id="student-supervision">Student supervision</h3>
<section id="master-students" class="level4">
<h4 class="anchored" data-anchor-id="master-students">Master students</h4>
<ul>
<li>Piotr Buczek, 2025, MSc Quantitative Finance
<ul>
<li><em>Distributional prediction of insurance claims using deep transformation models</em></li>
<li>WU Vienna</li>
</ul></li>
<li>Margarida Ramos Silva, 2025, MSc Quantitative Finance
<ul>
<li><em>Nonparametric Conditional Independence Tests: Discovering Important Predictors of Credit Default</em></li>
<li>WU Vienna</li>
</ul></li>
<li>Holly Vuarnoz, 2025, MSc Biostatistics
<ul>
<li><a href="https://www.biostat.uzh.ch/master-theses"><em>Toward causality in mobile sensing for depression: Modeling time-varying behavior with functional data</em></a></li>
<li>University of Zurich, co-supervised with Torsten Hothorn</li>
</ul></li>
<li>Andrea Staub, 2025, MSc Biostatistics
<ul>
<li><a href="https://www.biostat.uzh.ch/master-theses"><em>Selecting data modalities for clinical prediction models using deep neural networks</em></a></li>
<li>University of Zurich, co-supervised with Beate Sick</li>
</ul></li>
<li>Nicola Stella, 2024, MSc Statistics
<ul>
<li><a href="https://math.ethz.ch/sfs/education/master-theses.html"><em>Using anchor regression for out-of-distribution generalization of widely used risk prediction models </em></a></li>
<li>ETH Zurich, co-supervised with Olga Demler and Peter Bühlmann</li>
</ul></li>
<li>Andrea Götschi, 2022, MSc Biostatistics
<ul>
<li><a href="https://www.biostat.uzh.ch/master-theses"><em>Applications of ordinal neural network transformation models</em></a></li>
<li>University of Zurich, co-supervised with Beate Sick</li>
</ul></li>
<li>Maic Rakitta, 2021, MSc Statistics
<ul>
<li><a href="https://github.com/rakittam/glare"><em>Generalized linear anchor regression</em></a></li>
<li>ETH Zurich, co-supervised with Peter Bühlmann</li>
</ul></li>
<li>Tran Xuan Thanh Elsener, 2021, MSc Biostatistics
<ul>
<li><a href="https://www.biostat.uzh.ch/master-theses"><em>Hyperparameter tuning in linear anchor regression</em></a></li>
<li>University of Zurich, co-supervised with Beate Sick</li>
</ul></li>
</ul>
</section>
<section id="bachelor-students" class="level4">
<h4 class="anchored" data-anchor-id="bachelor-students">Bachelor students</h4>
<p>I am currently supervising 2 Bachelor students at WU Vienna. Previously, I have supervised 7 Bachelor students at WU Vienna in Data Science or Statistics.</p>
</section>
</section>
<section id="teaching-experience-1" class="level3">
<h3 class="anchored" data-anchor-id="teaching-experience-1">Teaching experience</h3>
<ul>
<li><p>STA406 Generalized Regression (UZH, FS19/20/21, lecturer/teaching assistant)</p></li>
<li><p>STA408 Statistical methods in epidemiology (UZH, SS20, lecturer/teaching assistant)</p></li>
<li><p>STA402/MAT924 Likelihood inference (UZH, FS19, teaching assistant)</p></li>
<li><p>Physical chemistry for molecular medicine (ALU, SS18, teaching assistant)</p></li>
</ul>
</section>
</section>
<section id="reviewing" class="level2">
<h2 class="anchored" data-anchor-id="reviewing">Reviewing</h2>
<section id="editorial-responsibilities" class="level3">
<h3 class="anchored" data-anchor-id="editorial-responsibilities">Editorial responsibilities</h3>
<ul>
<li><p>10/2025–present: Associate Editor for the <a href="https://www.jstatsoft.org/about/editorialTeam">Journal of Statistical Software</a></p></li>
<li><p>03/2025–present: Replication Editor for the <a href="https://www.jstatsoft.org/about/editorialTeam">Journal of Statistical Software</a></p></li>
<li><p>03/2024–present: Associate Editor for <a href="https://link.springer.com/journal/10182/editors">Advances in Statistical Analysis</a></p></li>
<li><p>09/2022–03/2025: Junior Editor for the <a href="https://www.jstatsoft.org/about/editorialTeam">Journal of Statistical Software</a></p></li>
</ul>
</section>
<section id="peer-review" class="level3">
<h3 class="anchored" data-anchor-id="peer-review">Peer review</h3>
<section id="journals" class="level4">
<h4 class="anchored" data-anchor-id="journals">Journals</h4>
<ul>
<li>Advances in Statistical Analysis</li>
<li>Electronic Journal of Statistics</li>
<li>Journal of Big Data</li>
<li>Journal of Computational and Graphical Statistics</li>
<li>Journal of Consumer Behaviour</li>
<li>Journal of the Royal Statistical Society, Series B (Statistical Methodology)</li>
<li>Knowledge and Information Systems</li>
<li>R Journal</li>
<li>Statistical Methods in Medical Research</li>
</ul>
</section>
<section id="conferences" class="level4">
<h4 class="anchored" data-anchor-id="conferences">Conferences</h4>
<ul>
<li><a href="http://aistats.org/aistats2026/">AISTATS 2026</a>, <a href="http://aistats.org/aistats2025/">AISTATS 2025</a>, <a href="http://aistats.org/aistats2024/">AISTATS 2024</a></li>
<li><a href="https://www.auai.org/uai2025/">UAI 2025</a>, <a href="https://www.auai.org/uai2024/">UAI 2024</a></li>
</ul>
</section>
</section>
</section>
<section id="awards-and-scholarships" class="level2">
<h2 class="anchored" data-anchor-id="awards-and-scholarships">Awards and Scholarships</h2>
<ul>
<li>12/2022: Postdoc.Mobility grant (Swiss National Science Foundation)
<ul>
<li>Title: <a href="https://data.snf.ch/grants/grant/214457">Distributional Instrumental Variables Estimation (DIVE)</a></li>
</ul></li>
<li>12/2021: GRC Travel Grant (University of Copenhagen, Host: Prof.&nbsp;Jonas Peters)</li>
<li>2015–2020: Full scholarship of the Academic Foundation of the German People</li>
<li>2015–2020: Full scholarship of the Lion’s Club Viersen e.V.</li>
</ul>
</section>
<section id="languages" class="level2">
<h2 class="anchored" data-anchor-id="languages">Languages</h2>
<ul>
<li>English (Full professional proficiency)</li>
<li>German (Native proficiency)</li>
<li>Mandarin Chinese (HSK 4)</li>
<li>French (Basic knowledge)</li>
</ul>
</section>
<section id="software" class="level2">
<h2 class="anchored" data-anchor-id="software">Software</h2>
<ul>
<li>Programming languages
<ul>
<li>R, Bash, Julia, Python</li>
</ul></li>
<li>Miscellaneous
<ul>
<li>Git, SVN, make, LaTeX, knitr, Docker</li>
</ul></li>
</ul>
<p>For more software-related projects, check out <a href="https://github.com/LucasKook">my GitHub profile</a>. I am co-maintainer of the <a href="https://CRAN.R-project.org/view=MachineLearning">CRAN Task View on Machine Learning &amp; Statistical Learning</a>.</p>
<section id="maintained-r-packages" class="level3">
<h3 class="anchored" data-anchor-id="maintained-r-packages">Maintained R packages</h3>
<p>Kook L, Mogensen SW (2026). “<code>glip</code>: Graph Learning via Integer Programming.” R package available <a href="https://github.com/LucasKook/glip.git">on GitHub</a>.</p>
<p>Kook L (2024). “<code>comets</code>: Covariance Measure Tests for Conditional Independence.” R package available <a href="https://github.com/LucasKook/comets.git">on GitHub</a> and <a href="https://CRAN.R-project.org/package=comets">CRAN</a>.</p>
<p>Kook L (2023). “<code>tramicp</code>: Model-Based Causal Feature Selection for General Response Types.” R package available <a href="https://github.com/LucasKook/tramicp.git">on GitHub</a> and <a href="https://CRAN.R-project.org/package=tramicp">on CRAN</a>.</p>
<p>Kook L, Baumann PFM, Rügamer D (2022). “<code>deeptrafo</code>: Fitting Deep Conditional Transformation Models.” R package available <a href="http://cran.r-project.org/package=deeptrafo">on CRAN</a>.</p>
<p>Kook L (2022). “<code>tramvs</code>: Optimal Subset Selection for Transformation Models.” R package available <a href="https://cran.r-project.org/package=tramvs">on CRAN</a>.</p>
<p>Kook L, Tamasi B, Siegfried S, Pawel S, Hothorn T (2021). “<code>tramnet</code>: Penalized Transformation Models.” R package available <a href="http://cran.r-project.org/package=tramnet">on CRAN</a>.</p>
</section>
<section id="contributed-r-packages" class="level3">
<h3 class="anchored" data-anchor-id="contributed-r-packages">Contributed R packages</h3>
<p>Hothorn T, Barbanti L, Siegfried S, Kook L (2025). “<code>tram</code>: Transformation Models.” R package available on <a href="https://CRAN.R-project.org/package=tram">CRAN</a>.</p>
<p>Rügamer D, Pfisterer F, Kook L, Baumann P, Kolb C (2022). “<code>deepregression</code>: Fitting Deep Distributional Regressions.” R package available <a href="http://cran.r-project.org/package=deepregression">on CRAN</a>.</p>
<p>Kook L, Wolski W (2021). “<code>prora</code>: ORA, sigORA and GSEA Functionalities for Proteomic Data.” R package available <a href="https://protviz.github.io/prora/index.html">on GitHub</a>.</p>
</section>
<section id="contributed-python-packages" class="level3">
<h3 class="anchored" data-anchor-id="contributed-python-packages">Contributed Python packages</h3>
<p>Huang S, Kook L (2024). “<code>pycomets</code>: Covariance Measure Tests for Conditional Independence.” Python package available <a href="https://github.com/shimenghuang/pycomets.git">on GitHub</a>.</p>
</section>
</section>
<section id="publications" class="level2">
<h2 class="anchored" data-anchor-id="publications">Publications</h2>
<p>My publications are also indexed on <a href="https://scholar.google.com/citations?hl=en&amp;user=mkbueQkAAAAJ&amp;view_op=list_works&amp;sortby=pubdate&amp;inst=13856427432203950092">my Google Scholar profile</a>.</p>
<section id="preprints" class="level3">
<h3 class="anchored" data-anchor-id="preprints">Preprints</h3>
<p><strong>Kook</strong> L, Mogensen SW (2026). “Exact Graph Learning via Integer Programming.” <em>arXiv preprint arXiv:2601.20589</em>. <a href="https://doi.org/10.48550/arXiv.2601.20589">doi</a>, <a href="https://arxiv.org/pdf/2601.20589.pdf">pdf</a>, <a href="https://github.com/LucasKook/glip">GitHub</a>.</p>
<p>Bajons R, <strong>Kook</strong> L (2025). “Rethinking player evaluation in sports: Goals above expectation and beyond.” <em>arXiv preprint arXiv:2509.20083</em>. <a href="https://doi.org/10.48550/arXiv.2509.20083">doi</a>, <a href="https://arxiv.org/pdf/2509.20083.pdf">pdf</a>, <a href="https://github.com/Rob2208/rGAX_and_beyond">GitHub</a>.</p>
<p><strong>Kook</strong> L, Götschi A, Baumann PF, Hothorn T, Sick B (2022). “Deep interpretable ensembles.” <em>arXiv preprint arXiv:2205.12729</em>. <a href="https://doi.org/10.48550/arXiv.2205.12729">doi</a>, <a href="https://arxiv.org/pdf/2205.12729.pdf">pdf</a>, <a href="https://github.com/neural-structured-additive-learning/deeptrafo.git">Software</a>.</p>
</section>
<section id="methodology" class="level3">
<h3 class="anchored" data-anchor-id="methodology">Methodology</h3>
<p>Doohan J, <strong>Kook</strong> L, Burke K (2026). “Comparison of generalised additive models and neural networks in applications: A systematic review.” <em>Expert Systems with Applications</em>. <a href="https://doi.org/10.1016/j.eswa.2025.131082">doi</a>.</p>
<p><strong>Kook</strong> L, Pfister N (2025). “Instrumental Variable Estimation of Distributional Causal Effects.” <em>Electronic Journal of Statistics</em>, <em>19</em>(2), 5249-5288. <a href="https://doi.org/10.1214/25-EJS2460">doi</a>, <a href="https://projecteuclid.org/journalArticle/Download?urlId=10.1214%2F25-EJS2460">pdf</a>, <a href="https://github.com/LucasKook/dive.git">GitHub</a>.</p>
<p>Campanella G, Häggström I, Kook L, Hothorn T, Fuchs TJ (2025). “A Flexible Deep Learning Framework for&nbsp;Survival Analysis with&nbsp;Medical Data.” In Gee JC, Alexander DC, Hong J, Iglesias JE, Sudre CH, Venkataraman A, Golland P, Kim JH, Park J (eds.), <em>Medical Image Computing and Computer Assisted Intervention - MICCAI 2025</em>, 3-13. <a href="https://doi.org/10.1007/978-3-032-05182-0_1">doi</a>.</p>
<p><strong>Kook</strong> L (2025). “Falsifying causal models via nonparametric conditional independence testing.” In O’Sullivan D, Sweeney J, Moghaddam S, Burke K (eds.), <em>Proceedings of the 39th International Workshop on Statistical Modelling (IWSM)</em>, 229-233. <a href="./papers/kook-iwsm-2025.pdf">pdf</a>, <a href="https://github.com/LucasKook/fcm-iwsm.git">GitHub</a>.</p>
<p>Bajons R, <strong>Kook</strong> L (2025). “Machine learning based statistical inference in sports analytics.” In O’Sullivan D, Sweeney J, Moghaddam S, Burke K (eds.), <em>Proceedings of the 39th International Workshop on Statistical Modelling (IWSM)</em>, 218-222. <a href="./papers/bajons-kook-iwsm-2025.pdf">pdf</a>.</p>
<p><strong>Kook</strong> L, Saengkyongam S, Lundborg AR, Hothorn T, Peters J (2025). “Model-based causal feature selection for general response types.” <em>Journal of the American Statistical Association</em>, <em>120</em>(550), 1090-1101. <a href="https://doi.org/10.1080/01621459.2024.2395588">doi</a>, <a href="./papers/kook-et-al-jasa-2024.pdf">pdf</a>, <a href="https://github.com/LucasKook/tramicp.git">GitHub</a>, <a href="https://CRAN.R-project.org/package=tramicp">Software</a>.</p>
<p><strong>Kook</strong> L, Baumann PFM, Dürr O, Sick B, Rügamer D (2024). “Estimating Conditional Distributions with Neural Networks using R package deeptrafo.” <em>Journal of Statistical Software</em>, <em>111</em>(10), 1-36. <a href="https://doi.org/10.18637/jss.v111.i10">doi</a>, <a href="https://www.jstatsoft.org/index.php/jss/article/view/v111i10/4653">pdf</a>, <a href="https://github.com/neural-structured-additive-learning/deeptrafo.git">Software</a>.</p>
<p><strong>Kook</strong> L, Lundborg AR (2024). “Algorithm-agnostic significance testing in supervised learning with multimodal data.” <em>Briefings in Bioinformatics</em>, <em>25</em>(6). <a href="https://doi.org/10.1093/bib/bbae475">doi</a>, <a href="./papers/kook-lundborg-bib-2024.pdf">pdf</a>, <a href="https://github.com/LucasKook/comets.git">GitHub</a>, <a href="https://CRAN.R-project.org/package=comets">CRAN</a>.</p>
<p><strong>Kook</strong> L, Kolb C, Schiele P, Dold D, Arpogaus M, Fritz C, Baumann PFM, Kopper P, Pielok T, Dorigatti E, Rügamer D (2024). “How Inverse Conditional Flows Can Serve as a Substitute for Distributional Regression.” <em>UAI 2024</em>. <a href="https://doi.org/10.48550/arXiv.2405.05429">doi</a>.</p>
<p>Rügamer D, Kolb C, Weber T, <strong>Kook</strong> L, Nagler T (2024). “Generalizing Orthogonalization for Models with Non-linearities.” <em>ICML 2024</em>. <a href="https://doi.org/10.48550/arXiv.2405.02475">doi</a>.</p>
<p>Siegfried S, <strong>Kook</strong> L, Hothorn T (2023). “Distribution-Free Location-Scale Regression.” <em>The American Statistician</em>, 1-12. <a href="https://doi.org/10.1080/00031305.2023.2203177">doi</a>, <a href="./papers/siegfried-kook-hothorn-2023.pdf">pdf</a>, <a href="https://CRAN.R-project.org/package=tramvs">Software</a>.</p>
<p>Marmolejo-Ramos F, Tejo M, Brabec M, Kuzilek J, Joksimovic S, Kovanovic V, González J, Kneib T, Bühlmann P, <strong>Kook</strong> L, Briseño-Sánchez G, Ospina R (2022). “Distributional regression modeling via generalized additive models for location, scale, and shape: An overview through a data set from learning analytics.” <em>WIREs Data Mining and Knowledge Discovery</em>, e1479. <a href="https://doi.org/10.1002/widm.1479">doi</a>, <a href="./papers/marmolejo-et-al-2022.pdf">pdf</a>.</p>
<p>Pawel S, <strong>Kook</strong> L, Reeve K (2023). “Pitfalls and potentials in simulation studies: Questionable research practices in comparative simulation studies allow for spurious claims of superiority of any method.” <em>Biometrical Journal</em>. <a href="https://doi.org/10.1002/bimj.202200091">doi</a>, <a href="./papers/pawel-kook-reeve-2023.pdf">pdf</a>, <a href="https://github.com/LucasKook/ainet.git">Software</a>.</p>
<p>Rügamer D, Kolb C, Fritz C, Pfisterer F, Kopper P, Bischl B, Shen R, Bukas C, Barros de Andrade e Sousa L, Thalmeier D, Baumann PFM, <strong>Kook</strong> L, Klein N, Müller CL (2023). “deepregression: A Flexible Neural Network Framework for Semi-Structured Deep Distributional Regression.” <em>Journal of Statistical Software</em>, <em>105</em>(1), 1–31. <a href="https://doi.org/10.18637/jss.v105.i02">doi</a>, <a href="https://www.jstatsoft.org/index.php/jss/article/view/v105i02/4412">pdf</a>, <a href="https://github.com/neural-structured-additive-learning/deepregression.git">Software</a>.</p>
<p>Herzog L, <strong>Kook</strong> L, Götschi A, Petermann K, Hänsel M, Hamann J, Wegener S, Sick B (2022). “Deep conditional transformation models for functional outcome prediction after acute ischemic stroke.” <em>Biometrial Journal</em>. <a href="https://doi.org/10.1002/bimj.202100379">doi</a>, <a href="./papers/herzog-kook-et-al-2022.pdf">pdf</a>, <a href="https://github.com/LucasKook/dtm-usz-stroke.git">GitHub</a>.</p>
<p><strong>Kook</strong> L, Sick B, Bühlmann P (2022). “Distributional Anchor Regression.” <em>Statistics and Computing</em>, <em>32</em>(3), 39. <a href="https://doi.org/10.1007/s11222-022-10097-z">doi</a>, <a href="./papers/kook-sick-buehlmann-2022.pdf">pdf</a>, <a href="https://github.com/LucasKook/distributional-anchor-regression.git">GitHub</a> , <a href="https://github.com/LucasKook/dare.git">Software</a>.</p>
<p><strong>Kook</strong> L, Herzog L, Hothorn T, Dürr O, Sick B (2022). “Deep and Interpretable Regression Models for Ordinal Outcomes.” <em>Pattern Recognition</em>. <a href="https://doi.org/10.1016/j.patcog.2021.108263">doi</a>, <a href="./papers/kook-herzog-et-al-2022.pdf">pdf</a>, <a href="https://github.com/neural-structured-additive-learning/deeptrafo.git">Software</a>.</p>
<p><strong>Kook</strong> L, Hothorn T (2021). “Regularized Transformation Models: The tramnet Package.” <em>The R Journal</em>. <a href="https://doi.org/10.32614/RJ-2021-054">doi</a>, <a href="./papers/kook-hothorn-2021.pdf">pdf</a>, <a href="https://CRAN.R-project.org/package=tramnet">Software</a>.</p>
</section>
<section id="biomedicine" class="level3">
<h3 class="anchored" data-anchor-id="biomedicine">Biomedicine</h3>
<p>Xie JS, Zaslavsky K, Chaban Y, Lusterio A, Kaur H, Motekalem Y, Zeraatkar D, Popovic MM, Althaus K, Malbin B, Nolte CH, Scheitz JF, Pettersen JA, Leker RR, Kim J, Woo SJ, Chen C, Feltgen N, Khrlobyan M, Sangha N, Jurado EA, Arnold M, Mattle H, Heldner MR, Nedelmann MG, Cordonnier C, Spitzer MS, Poli S, Hametner C, Baumgartner P, Wegener S, <strong>Kook</strong> L, Shahjouei S, Dumitrascu OM, Margolin E, Michel P (2025). “Early Thrombolysis and Outcomes in Central Retinal Artery Occlusion: An Individual Participant Data Meta-Analysis.” <em>Stroke</em>. <a href="https://doi.org/10.1161/STROKEAHA.124.049955">doi</a>.</p>
<p>Baumgartner P, <strong>Kook</strong> L, Altersberger VL, Gensicke H, Ardila-Jurado E, Kägi G, Salerno A, Michel P, Gopisingh KM, Nederkoorn PJ, Scheitz JF, Nolte CH, Heldner MR, Arnold M, Cordonnier C, Schiava LD, Hametner C, Ringleb PA, Leker RR, Jubran H, Luft AR, Engelter ST, Wegener S, the Thrombolysis F (2023). “Safety and effectiveness of IV Thrombolysis in retinal artery occlusion: A multicenter retrospective cohort study.” <em>European Stroke Journal</em>. <a href="https://doi.org/10.1177/23969873231185895">doi</a>, <a href="./papers/baumgartner-et-al-2023.pdf">pdf</a>.</p>
<p>Herzog L, <strong>Kook</strong> L, Hamann J, Globas C, Heldner MR, Seiffge D, Antonenko K, Dobrocky T, Panos L, Kaesmacher J, Fischer U, Gralla J, Arnold M, Wiest R, Luft AR, Sick B, Wegener S (2023). “Deep Learning Versus Neurologists: Functional Outcome Prediction in LVO Stroke Patients Undergoing Mechanical Thrombectomy.” <em>Stroke</em>. <a href="https://doi.org/10.1161/STROKEAHA.123.042496">doi</a>, <a href="./papers/herzog-et-al-2023.pdf">pdf</a>.</p>
<p>Schmidt MT, Studer M, Kunz A, Studer S, Bonvini JM, Bueter M, <strong>Kook</strong> L, Haile SR, Pregernig A, Beck-Schimmer B, Schläpfer M (2023). “There is no evidence that carbon dioxide-enriched oxygen before apnea affects the time to arterial desaturation, but it might improve cerebral oxygenation in anesthetized obese patients: a single-blinded randomized crossover trial.” <em>BMC Anesthesiology</em>, <em>23</em>(1), 41. <a href="https://doi.org/10.1186/s12871-023-01982-9">doi</a>, <a href="./papers/schmidt-et-al-2023.pdf">pdf</a>.</p>
<p>Fröhlich K, Brombacher E, Fahrner M, Vogele D, <strong>Kook</strong> L, Pinter N, Bronsert P, Timme-Bronsert S, Schmidt A, Bärenfaller K, Kreutz C, Schilling O (2022). “Benchmarking of analysis strategies for data-independent acquisition proteomics using a large-scale dataset comprising inter-patient heterogeneity.” <em>Nature Communications</em>, <em>13</em>(1), 2622. <a href="https://doi.org/10.1038/s41467-022-30094-0">doi</a>, <a href="./papers/froehlich-et-al-2023.pdf">pdf</a>.</p>
<p>Fahrner M, <strong>Kook</strong> L, Fröhlich K, Biniossek ML, Schilling O (2021). “A Systematic Evaluation of Semispecific Peptide Search Parameter Enables Identification of Previously Undescribed N-Terminal Peptides and Conserved Proteolytic Processing in Cancer Cell Lines.” <em>Proteomes</em>, <em>9</em>(2). <a href="https://doi.org/10.3390/proteomes9020026">doi</a>, <a href="./papers/fahrner-et-al-2021.pdf">pdf</a>.</p>
<p>Ohmer M, Tzivelekidis T, Niedenführ N, Volceanov-Hahn L, Barth S, Vier J, Börries M, Busch H, <strong>Kook</strong> L, Biniossek ML, Schilling O, Kirschnek S, Häcker G (2019). “Infection of HeLa cells with Chlamydia trachomatis inhibits protein synthesis and causes multiple changes to host cell pathways.” <em>Cellular microbiology</em>, <em>21</em>(4), e12993. <a href="https://doi.org/10.1111/cmi.12993">doi</a>, <a href="./papers/ohmer-et-al-2018.pdf">pdf</a>.</p>
<p>Drendel V, Heckelmann B, Schell C, <strong>Kook</strong> L, Biniossek ML, Werner M, Jilg CA, Schilling O (2018). “Proteomic distinction of renal oncocytomas and chromophobe renal cell carcinomas.” <em>Clinical proteomics</em>, <em>15</em>(1), 25. <a href="https://doi.org/10.1186/s12014-018-9200-6">doi</a>, <a href="./papers/drendel-et-al-2018.pdf">pdf</a>.</p>
<p>Saporito-Magriñá CM, Musacco-Sebio RN, Andrieux G, <strong>Kook</strong> L, Orrego MT, Tuttolomondo MV, Desimone MF, Boerries M, Borner C, Repetto MG (2018). “Copper-induced cell death and the protective role of glutathione: the implication of impaired protein folding rather than oxidative stress.” <em>Metallomics</em>, <em>10</em>(12), 1743-1754. <a href="./papers/saporito-et-al-2018.pdf">pdf</a>.</p>
</section>
</section>
<section id="talks" class="level2">
<h2 class="anchored" data-anchor-id="talks">Talks</h2>
<p>Kook L (2025). “Falsifying causal models via nonparametric conditional independence testing.” IWSM, Limerick, Ireland, received the “Early-Career Presentation Award (Runner-up)”.</p>
<p>Kook L (2025). “Regression-based conditional independence testing in statistics and causal inference.” Invited talk (host: Pawel Morzywolek, online), University of Washington, USA.</p>
<p>Kook L (2024). “Instrumental variable estimation of distributional causal effects.” Invited talk, ICSDS, Nice, France.</p>
<p>Kook L (2024). “Instrumental variable estimation of distributional causal effects.” Invited talk, University of Limerick, Ireland.</p>
<p>Kook L (2024). “Instrumental variable estimation of distributional causal effects.” Invited talk (host: Stefan Feuerriegel), LMU Munich, Germany.</p>
<p> L (2024). “Algorithm-agnostic inference with multimodal data.” Invited talk, Bernoulli-ims 11th World Congress in Probability and Statistics, Bochum, Germany.</p>
<p>Kook L (2023). “Model-based causal feature selection for general response types.” Invited talk, UZH Econometrics Seminar, Zurich, Switzerland.</p>
<p>Kook L (2023). “Distributional regression with neural networks in R.” Invited talk, CMStatistics Berlin, Germany.</p>
<p>Kook L (2023). “Invariant causal prediction for non-additive noise models.” Contributed talk, EMS Warsaw, Poland.</p>
<p>Kook L (2023). “Invariant causal prediction for non-additive noise models.” Contributed talk, NORDSTAT Gothenburg, Sweden.</p>
<p>Kook L (2022). “Ensembling deep transformation models.” Invited talk, CMStatistics London, Great Britain.</p>
<p>Kook L, Herzog L, Dürr O, Hothorn T, Wegener S, Sick B (2021). “Interpretable effect estimates in semi-structured deep distributional regression.” ISCB’42, Lyon (online).</p>
<p>Kook L (2021). “Ordinal neural network transformation models.” Invited talk, LMU Munich (online).</p>
<p>Kook L, Herzog L, Grabner H, Wegener S, Sick B (2021). “Improving treatment decisions in acute stroke using deep-learning based risk analysis.” Freenovation Science Forum (online).</p>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/LucasKook\.github\.io\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>